### ISSUE TITLE
sprint production release 2021.12.20

### ISSUE THREAD

**Issue by harshad16** [thoth-station/thoth-application#2183]
Hello, Thoth-station!

This Issue would be used for the current sprint cycle production release.
By the end of the sprint cycle, we will consolidate the information of thoth-station components features upgrade and fixes in this issue.

**Comment by fridex** [comment]
*List available environments*

Starting this release, users can issue `thamos environments` to list available environments for the resolution. This allows users to ask which environments are available and can be used in Thoth's configuration file:

* https://github.com/thoth-station/user-api/pull/1529
* https://github.com/thoth-station/thamos/pull/961

**Comment by fridex** [comment]
**Warn if users use too lax versions in their requirements**

Resolver warns if users use too lax versions in their requirements. It is a good practice to specify desired versions in the requirements file that are tested and expected to work, not to introduce overplanning issues.

* https://github.com/thoth-station/adviser/pull/2204

**Comment by fridex** [comment]
**Links to Search UI**

Thanks to [Thoth Search](https://github.com/thoth-station/thoth-search) project, people can browse the resolver results in a web browser. The link to Thoth Search can be found in each resolver result (might not immediately work with the current release as Thoth Search is to be deployed).

* https://github.com/thoth-station/adviser/pull/2205

**Comment by mayaCostantini** [comment]
**Create an example application in `cli-examples` to demonstrate Thamos dependency resolution**

The `game_of_life.py` simple application allows users to understand how Thoth detects vulnerabilities in an application software stack and prevents the import of vulnerable packages when put on the `security` setting.

- https://github.com/thoth-station/cli-examples/issues/21
- https://github.com/thoth-station/cli-examples/pull/23

**Comment by fridex** [comment]
**Warn if a package has no recent releases on PyPI**

We have added prescriptions that warn users if they use a package that has no recent releases on PyPI (180+ days). This might indicate that the project is unmaintained -- especially when combined together with other prescriptions that we have about projects (such as repo activity). Prescriptions created are automatically updated by the prescriptions-refresh-job run periodically in deployment.

* https://github.com/thoth-station/prescriptions/pull/18634
* https://github.com/thoth-station/prescriptions-refresh-job/pull/68

**Comment by fridex** [comment]
**Warnings produced based on PyPI maintainers**

Newly, resolver warns about using projects that can have suspicious behaviour based on statistics computed from PyPI maintainers info:

* resolver warns if a package has not enough maintainers on PyPI
* resolver warns if a package is maintained by maintainers that maintain a small number of projects on PyPI
* resolver warns if all package maintainers for a package hosted on PyPI have joined PyPI just recently

Related:
* https://github.com/thoth-station/prescriptions-refresh-job/pull/78

**Comment by mayaCostantini** [comment]
**Prescriptions produced to inform users on the number of downloads for a given package**

New prescriptions inform users on the number of downloads for a given package on PyPI and on the number of downloads for each package version in the last 180 days according to data from the [PyPI downloads BigQuery dataset](https://console.cloud.google.com/bigquery?p=bigquery-public-data&d=pypi&page=dataset&project=data-watch-333811). Prescriptions for a project include the most downloaded project version, and prescriptions for each project version score the popularity of the given version. 

- https://github.com/thoth-station/prescriptions-refresh-job/issues/61
- https://github.com/thoth-station/prescriptions-refresh-job/pull/65

**Comment by fridex** [comment]
**Installed Python artifact size**

Yet another new set of prescriptions show to users the downloaded artifact size that will be eventually installed. Note that this is just an estimate (upper limit) as the actual artifact installed depends on decisions done by pip during the resolution for the target runtime environment.

* https://github.com/thoth-station/prescriptions-refresh-job/pull/80
* https://github.com/thoth-station/prescriptions/pull/18636

**Comment by fridex** [comment]
**Size of container images used**

Similar to Python artifact size, users get notified about container image size if Thoth containerized environments are used.

* https://github.com/thoth-station/prescriptions-refresh-job/pull/84
* https://github.com/thoth-station/prescriptions-refresh-job/issues/81
* https://github.com/thoth-station/prescriptions/pull/18638

**Comment by Gkrumbach07** [comment]
**Migrated thoth-search repo over to Thoth Station**

The Thoth Search app was originally under @Gkrumbach07 ownership but has been transferred over to Thoth Station. Bots have been added and pre-commits have been configured.  Also there has been changes in the CI/CD GitHub workflow that now check for `ESlint` errors and `prettier` formatting issues. The app is deployed at https://thoth-station.ninja/thoth-search/

- https://github.com/thoth-station/thoth-application/issues/2034
- https://github.com/thoth-station/thoth-search/issues/29
- thoth-station/thoth-search/pull/26

**Comment by fridex** [comment]
**Filtering Python packages used during the resolution based on `index` labels supplied to the resolver**

Users can now use labels to distinguish Python package indexes that should be used during the resolution process. This is an addition to [the strict index configuration](https://thoth-station.ninja/docs/developers/adviser/experimental_features.html).

* https://github.com/thoth-station/prescriptions/pull/18644
* https://github.com/thoth-station/adviser/pull/2216

An example can be an advise for a Python package that should not use packages hosted on PyPI nor PyTorch CUDA 11.1 index but use all the others monitored by Thoth:

```
thamos advise --labels "index=no-pypi,index=no-pytorch-cu111"
```

or use solely packages hosted on Pulp instance on the Operate First environment:

```
thamos advise --labels "index=pulp"
```

**Comment by fridex** [comment]
**Labels in Thoth's configuration file**

To simplify label handling and to give users the ability to specify persistent labels, Thamos CLI accepts `labels` section in each runtime environment entry in `.thoth.yaml` file. Labels stated there are specific to the runtime environment used and can additionally configure how the resolution process should look like for the desired runtime environment. Labels are used to additionally select pipeline units that adjust the resolution process.

* https://github.com/thoth-station/thamos/pull/965

**Comment by fridex** [comment]
**Python package information stored in each container image produced by AICoE-CI**

AICoE-CI now propagates information about packages used in a form of `Pipfile`, `Pipfile.lock` stored in `/opt/aicoe-ci` thanks to @harshad16. This enabled propagating this information in container image analyses done in Thoth's package-extract which enables to introspect what Python packages from which indexes are used during the build.

* https://github.com/AICoE/aicoe-ci/pull/154
* https://github.com/thoth-station/package-extract/pull/459

**Comment by harshad16** [comment]
we have completed the release of 2021.12.20 :tada: :confetti_ball: :partying_face: 

### Features

**List available environments**

Starting this release, users can issue `thamos environments` to list available environments for the resolution. This allows users to ask which environments are available and can be used in Thoth's configuration file:

* https://github.com/thoth-station/user-api/pull/1529
* https://github.com/thoth-station/thamos/pull/961

**Warn if users use too lax versions in their requirements**

Resolver warns if users use too lax versions in their requirements. It is a good practice to specify desired versions in the requirements file that are tested and expected to work, not to introduce overplanning issues.

* https://github.com/thoth-station/adviser/pull/2204
* https://github.com/thoth-station/thoth-station.github.io/pull/235

**Links to Search UI**

Thanks to [Thoth Search](https://github.com/thoth-station/thoth-search) project, people can browse the resolver results in a web browser. The link to Thoth Search can be found in each resolver result (might not immediately work with the current release as Thoth Search is to be deployed).

* https://github.com/thoth-station/adviser/pull/2205

The resolver also creates a link for each resolved package to the corresponding Thoth Search UI package entry:

* https://github.com/thoth-station/adviser/pull/2214/

**Create an example application in `cli-examples` to demonstrate Thamos dependency resolution**

The `game_of_life.py` simple application allows users to understand how Thoth detects vulnerabilities in an application software stack and prevents the import of vulnerable packages when put on the `security` setting.

- https://github.com/thoth-station/cli-examples/issues/21
- https://github.com/thoth-station/cli-examples/pull/23

**Warn if a package has no recent releases on PyPI**

We have added prescriptions that warn users if they use a package that has no recent releases on PyPI (180+ days). This might indicate that the project is unmaintained -- especially when combined together with other prescriptions that we have about projects (such as repo activity). Prescriptions created are automatically updated by the prescriptions-refresh-job run periodically in deployment.

* https://github.com/thoth-station/prescriptions/pull/18634
* https://github.com/thoth-station/prescriptions-refresh-job/pull/68

**Warnings produced based on PyPI maintainers**

Newly, resolver warns about using projects that can have suspicious behaviour based on statistics computed from PyPI maintainers info:

* resolver warns if a package has not enough maintainers on PyPI
  * [example](https://github.com/thoth-station/prescriptions/blob/62ca4ea500e4f7987d91be4d54b7b579765609c9/prescriptions/2c_/2captcha-python/pypi_project_maintainers.yaml)
* resolver warns if a package is maintained by maintainers that maintain a small number of projects on PyPI
  * [example](https://github.com/thoth-station/prescriptions/blob/62ca4ea500e4f7987d91be4d54b7b579765609c9/prescriptions/2c_/2captcha-python/pypi_maintainers.yaml)
* resolver warns if all package maintainers for a package hosted on PyPI have joined PyPI just recently
  * [example](https://github.com/thoth-station/prescriptions/blob/62ca4ea500e4f7987d91be4d54b7b579765609c9/prescriptions/at_/atheris/pypi_maintainers_joined.yaml
)

Related:
* https://github.com/thoth-station/prescriptions-refresh-job/pull/78
* https://github.com/thoth-station/prescriptions/commit/62ca4ea500e4f7987d91be4d54b7b579765609c9 (the change might be too large to browse all the changes, check thoth-station/prescriptions repo for prescriptions available)

**Prescriptions produced to inform users on the number of downloads for a given package**

New prescriptions inform users on the number of downloads for a given package on PyPI and on the number of downloads for each package version in the last 180 days according to data from the [PyPI downloads BigQuery dataset](https://console.cloud.google.com/bigquery?p=bigquery-public-data&d=pypi&page=dataset&project=data-watch-333811). Prescriptions for a project include the most downloaded project version, and prescriptions for each project version score the popularity of the given version. 

- https://github.com/thoth-station/prescriptions-refresh-job/issues/61
- https://github.com/thoth-station/prescriptions-refresh-job/pull/65

**Installed Python artifact size**

Yet another new set of prescriptions show to users the downloaded artifact size that will be eventually installed. Note that this is just an estimate (upper limit) as the actual artifact installed depends on decisions done by pip during the resolution for the target runtime environment. Reports show information only for artifacts that have 3MiB+.

* https://github.com/thoth-station/prescriptions-refresh-job/pull/80
* https://github.com/thoth-station/prescriptions/pull/18637
* https://github.com/thoth-station/prescriptions/pull/18643
* https://github.com/thoth-station/prescriptions-refresh-job/pull/89

**Size of container images used**

Similar to Python artifact size, users get notified about container image size if Thoth containerized environments are used. The container image size is automatically recomputed in prescriptions-refresh-job each time there is a versioned release done by AICoE-CI.

* https://github.com/thoth-station/prescriptions-refresh-job/pull/84
* https://github.com/thoth-station/prescriptions-refresh-job/issues/81
* https://github.com/thoth-station/prescriptions/pull/18638

**Migrated thoth-search repo over to Thoth Station**

The Thoth Search app has been transferred over to Thoth Station. Bots have been added and pre-commits have been configured.  Also there has been changes in the CI/CD GitHub workflow that now check for `ESlint` errors and `prettier` formatting issues. The app is deployed at https://thoth-station.ninja/thoth-search/

- https://github.com/thoth-station/thoth-application/issues/2034
- https://github.com/thoth-station/thoth-search/issues/29
- thoth-station/thoth-search/pull/26

**Filtering Python packages used during the resolution based on `index` labels supplied to the resolver**

Users can now use labels to distinguish Python package indexes that should be used during the resolution process. This is an addition to [the strict index configuration](https://thoth-station.ninja/docs/developers/adviser/experimental_features.html).

* https://github.com/thoth-station/prescriptions/pull/18644
* https://github.com/thoth-station/adviser/pull/2216

An example can be an advise for a Python application that should not use packages hosted on PyPI nor PyTorch CUDA 11.1 index but use all the others monitored by Thoth:

```
thamos advise --labels "pypi-index=disabled,pytorch-cu111-index=disabled"
```

or use solely packages hosted on Pulp indexes on the Operate First instance:

```
thamos advise --labels "opf-pulp-indexes=solely"
```

The full listing of labels with values available:

```
- pytorch-cpu-index: {disabled,solely}
- pytorch-cu111-index: {disabled,solely}
- aicoe-avx2-tf-index: {disabled,solely}
- pypi-index: {disabled,solely}
- opf-pulp-indexes: {disabled,solely}
```
**Labels in Thoth's configuration file**

To simplify label handling and to give users the ability to specify persistent labels, Thamos CLI accepts `labels` section in each runtime environment entry in `.thoth.yaml` file. Labels stated there are specific to the runtime environment used and can additionally configure how the resolution process should look like for the desired runtime environment. Labels are used to additionally select pipeline units that adjust the resolution process.

* https://github.com/thoth-station/thamos/pull/965
* https://github.com/thoth-station/thamos/issues/963

Available since `thamos>=1.22.0`

**Python package information stored in each container image produced by AICoE-CI**

AICoE-CI now propagates information about packages used in a form of `Pipfile`, `Pipfile.lock` stored in `/opt/aicoe-ci`. This enabled propagating this information in container image analyses done in Thoth's package-extract which enables to introspect what Python packages from which indexes are installed during the application build.

* https://github.com/AICoE/aicoe-ci/pull/154
* https://github.com/thoth-station/package-extract/pull/459

### Component Updates

- [x] adviser: v0.48.0
  - https://github.com/thoth-station/adviser/pull/2161


Thanks for the amazing work everyone. :100:

**Comment by harshad16** [comment]
Thanks, everyone 
Have a wonderful holidays :snowman: :snowman_with_snow: :snowflake: :christmas_tree: :santa: :gifts :gift_heart:
